subs/pantry/src/Pantry.hs

```diff
@@ -626,7 +628,35 @@ loadFromURL
   => Text -- ^ url
   -> Maybe BlobKey
   -> RIO env ByteString
-loadFromURL = undefined
+loadFromURL url Nothing = do
+  mcached <- withStorage $ loadURLBlob url
+  case mcached of
+    Just bs -> return bs
+    Nothing -> loadWithCheck url $ \_ -> return ()
+loadFromURL url (Just bkey@(BlobKey sha size)) = do
+  mcached <- withStorage $ loadBlob bkey
+  case mcached of
+    Just bs -> return bs
+    Nothing -> loadWithCheck url $ \bs -> do
+      let blobSha = mkStaticSHA256FromBytes bs
+          blobSize = FileSize $ fromIntegral $ B.length bs
+      when (blobSha /= sha || blobSize /= size) $
+        throwIO $ BadKey . utf8BuilderToText $
+          "Expected SHA/size : " <> display sha <> "/" <> display size <>
```

Comment:
Instead of including a `Text` here, the normal approach in the library would be to include the expected and actual `BlobKey`s in the data constructor, and doing the rendering in the `Show` instance.